custom_components/homewhiz/__init__.py

```diff
@@ -1,111 +1,114 @@
-
-    await async_get_custom_components(hass)
-    hass.config_entries.async_setup_platforms(entry, [Platform.SENSOR])
+    address = entry.unique_id
+    coordinator = hass.data.setdefault(DOMAIN, {})[
+        entry.entry_id
+    ] = HomewhizDataUpdateCoordinator(hass, address)
+
+    @callback
+    def connect(
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        hass.async_create_task(coordinator.connect())
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    entry.async_on_unload(
+        bluetooth.async_register_callback(
+            hass,
+            connect,
+            BluetoothCallbackMatcher(address=address),
+            bluetooth.BluetoothScanningMode.ACTIVE,
+        )
+    )
+
     return True
 
 
-class HomewhizDataUpdateCoordinator(DataUpdateCoordinator[WasherState]):
+class HomewhizDataUpdateCoordinator(DataUpdateCoordinator[WasherState | None]):
     def __init__(
         self,
         hass: HomeAssistant,
-        device: BLEDevice,
+        address: str,
     ) -> None:
-        self._device = device
-        self.client = BleakClient(device)
-        self.accumulator = MessageAccumulator()
+        self.address = address
+        self._accumulator = MessageAccumulator()
+        self._hass = hass
+        self._conn: BleakClient | None = None
+        self.device: BLEDevice | None = None
         super().__init__(hass, _LOGGER, name=DOMAIN)
 
     async def connect(self):
-        await self.connect_internal()
-        self.client.set_disconnected_callback(
-            lambda client: self.hass.create_task(self.reconnect(client))
+        self.device = bluetooth.async_ble_device_from_address(
+            self._hass, self.address, connectable=True
         )
-        await self.start_listening()
-        return True
-
-    async def start_listening(self):
-        await self.client.start_notify(
+        if not self.device:
+            raise Exception(f"Device not found for address {self.address}")
+        self._conn = await establish_connection(
+            client_class=BleakClient,
+            device=self.device,
+            disconnected_callback=lambda client: self.handle_disconnect(),
+            name=self.address,
+        )
+        if not self._conn.is_connected:
+            raise Exception(f"Can't connect to ${self.address}")
+        await self._conn.start_notify(
             "0000ac02-0000-1000-8000-00805f9b34fb",
             lambda sender, message: self.hass.create_task(
                 self.handle_notify(sender, message)
             ),
         )
-        await self.client.write_gatt_char(
+        await self._conn.write_gatt_char(
             "0000ac01-0000-1000-8000-00805f9b34fb",
             bytearray.fromhex("02 04 00 04 00 1a 01 03"),
             response=False,
         )
+        _LOGGER.info(f"[{self.address}] Successfully connected")
```

Comment:
Isn't hass connecting to the device and not the other way around? Anyway, lowercase "successfully" 😄

Reply:
I figured having the address in every log message could be helpful. Instead of writing "Successfully connected to address" I decided to use the address as a constant prefix. e.g 
`2022-12-02 00:42:26.424 INFO (MainThread) [custom_components.homewhiz] [D8FE97CF-4576-0553-0E50-87593872CF10] Successfully connected. RSSI: -79`

Reply:
Tell me if it's readable for you

Reply:
That makes sense, good idea!